// @dainprotocol/service-sdk/dist/plugins/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseDainPlugin = void 0;
/**
 * Base class for implementing DAIN plugins
 * Provides default implementations of required methods
 */
class BaseDainPlugin {
    /**
     * Get plugin data from a tool context
     * This simplifies accessing plugin data in a tool handler
     */
    getInputData(context) {
        return context?.extraData?.plugins?.[this.id];
    }
    /**
     * Create a standardized response for this plugin
     * This helps format plugin data consistently in tool responses
     */
    createResponse(data) {
        return data;
    }
    /**
     * Get processed plugin data from a tool response
     */
    getResponseData(response) {
        return response?.plugins?.[this.id];
    }
    /**
     * Process input data before sending to service
     * This is called on the client side before making a tool call
     */
    async processInputClient(input) {
        // Default implementation doesn't modify input
        return input;
    }
    /**
     * Process output data when receiving a tool response
     * This is called on the client side after receiving a tool response
     */
    async processOutputClient(output) {
        // Default implementation doesn't modify output
        return output;
    }
    /**
     * Process input data when receiving a tool call
     * This is called on the service side before processing a tool request
     */
    async processInputService(input) {
        // Default implementation doesn't modify input
        return input;
    }
    /**
     * Process output data when sending a tool response
     * This is called on the service side after processing a tool response
     */
    async processOutputService(output) {
        // Default implementation doesn't modify output
        return output;
    }
}
exports.BaseDainPlugin = BaseDainPlugin;
